feat: parsing and printing optimized

augi opened the pull request:
We are dealing with higher CPU usage in some of our applications and performance monitoring (using Flight Recorder) showed that the issue could be in `Sha256` decoding, which is used extensively in the applications.

Even the stacktrace from parsing is pretty deep:
![image](https://user-images.githubusercontent.com/273052/125648906-9c6ab5b5-a6b2-44a0-a1e6-37618d6ab7e2.png)

So I decided to rewrite the critical parts from _Scalish_ code to _hardcode_ optimized code (but still readable). The implementation is inspired by [code](https://github.com/scodec/scodec-bits/blob/main/core/shared/src/main/scala/scodec/bits/ByteVector.scala) [from scodec](https://github.com/scodec/scodec-bits/blob/main/core/shared/src/main/scala/scodec/bits/Bases.scala). I prefer to keep this library dependencies-less, so not to add the scodec as a dependency of this project.

Please note that `while` must be used because [optimized for-cycle for basic usages is still not a solved thing](https://github.com/scala/bug/issues/1338).

jendakol reviewed (commented):
I'm OK with this as I definitely do understand the problem 👍🏻 However, could you add some more tests? Test both ways, maybe add some tricky parts (like decoding string with unsupported chars etc.)...

augi commented:
This PR actually increased the code-coverage but I've just added few more tests (that work also as documentation), and I've also added comments to the methods as their behavior could be tricky.

jendakol commented on core/src/test/scala/com/avast/scala/hashes/PackageTest.scala:
```diff
@@ -0,0 +1,49 @@
+    ("001991FF", 5, None),
+    ("001991FF", 4, Some(Array(0, 25, 145, 255).map(_.toByte)))).foreach { case (input, expectedBytes, expectedResult) =>
+    it must s"return expected result from tryHex2bytes for input '$input' string and expecting $expectedBytes bytes" in {
+      tryHex2bytes(input, expectedBytes) match {
+        case Some(value) => value shouldBe expectedResult.get
+        case None => expectedResult shouldBe None
+      }
```
Wouldn't `tryHex2bytes(input, expectedBytes) shouldBe expectedResult` work better? 😉

augi replied:
I wouldn't because the _matchers_ are not able to handle `Array` equality if the array is included in an `Option` 😢 This is why I had to _expand_ the assertion code like this.

jendakol commented:
TBH while I read ot for the second time, I don't get much what is done there. On the other hand, it seems consistent with what _was_ done there until now so I hope it's ok 😄

augi commented:
Do you mean the comments? Yes, it also took me a while to realize what was the original behavior 😄 And as you wrote, the behavior remains the same after this PR.
